mcp_extension/trust/trust.go

```diff
@@ -0,0 +1,235 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package trust
+
+import (
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"html/template"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mark3labs/mcp-go/mcp"
+	"github.com/pkg/browser"
+	"github.com/rs/zerolog"
+	"github.com/snyk/go-application-framework/pkg/configuration"
+
+	"github.com/snyk/snyk-ls/mcp_extension/networking"
+)
+
+const (
+	TrustedFoldersConfigKey = "TRUSTED_FOLDERS"
+	DisableTrustFlag        = "disable-trust"
+)
+
+type FolderTrust struct {
+	logger *zerolog.Logger
+	config configuration.Configuration
+	mutex  sync.RWMutex
+}
+
+//go:embed trust.html
+var SnykTrustPage string
+
+func NewFolderTrust(logger *zerolog.Logger, config configuration.Configuration) *FolderTrust {
+	return &FolderTrust{
+		logger: logger,
+		config: config,
+	}
+}
+
+func normalizePath(folder string) string {
+	return path.Clean(folder)
+}
+
+func folderContains(folderPath string, path string) bool {
+	filePathSeparator := string(filepath.Separator)
+	cleanPath := normalizePath(path)
+	cleanFolderPath := normalizePath(folderPath)
+	if !strings.HasSuffix(cleanFolderPath, filePathSeparator) {
+		cleanFolderPath += filePathSeparator
+	}
+
+	// Check if the path is on a case-insensitive filesystem
+	if runtime.GOOS == "windows" {
+		cleanPath = strings.ToLower(cleanPath)
+		cleanFolderPath = strings.ToLower(cleanFolderPath)
+	}
+
+	return strings.HasPrefix(cleanPath, cleanFolderPath) ||
+		strings.HasPrefix(cleanPath+filePathSeparator, cleanFolderPath)
+}
+
+func (t *FolderTrust) IsFolderTrusted(folder string) bool {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+
+	for _, trustedFolder := range t.trustedFolders() {
+		if folderContains(trustedFolder, folder) {
+			return true
+		}
+	}
+	return false
+}
+
+func (t *FolderTrust) TrustedFolders() []string {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+
+	return t.trustedFolders()
+}
+
+func (t *FolderTrust) trustedFolders() []string {
+	rawRes := t.config.Get(TrustedFoldersConfigKey)
+	trustedFolders, ok := rawRes.([]string)
+	if !ok {
+		t.logger.Error().Msg("incorrect type stored for trusted folders")
+		return []string{}
+	}
+	return trustedFolders
+}
+
+func (t *FolderTrust) AddTrustedFolder(folder string) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	t.addTrustedFolder(folder)
+}
+
+func (t *FolderTrust) addTrustedFolder(folder string) {
+	trustedFolders := t.trustedFolders()
+	folder = normalizePath(folder)
+	if slices.Contains(trustedFolders, folder) {
```

Comment:
Hm. Not sure this is correct. Most likely you should loop (or use an iterator function or both) and use `FolderContains`.